File: DP-1-expressN/expressN.py
def combi(A, B):
    result = set()
    for a in A:
        for b in B:
            result |= {a*b}
            if b != 0:
                result |= {a//b}
            result |= {a+b}
            result |= {a-b}
    return result


def solution(N, number):
    d = [-1] * 10
    d[0] = {0}
    d[1] = {N}
    if number == N:
        return 1

    for i in range(2, 9):
        tmp = set()
        for j in range(1, i):
            tmp |= combi(d[j], d[i-j])
        tmp |= {int(str(N)*i)}
        d[i] = tmp
        if number in d[i]:
            return i
    return -1


N = 5
number = 26
print(solution(N, number))
'''
N	number	return
5	12  	4
2	11	    3
[{0}, {5}, {0, 1.0, 10, 55, 25}, {0, 2, 4.0, 5, 6.0, 5001, 11, 525, -50, 15.0, 275, 20, 30, 105, 555, -4.0, -20, 50, -5, 60, 125, 510, 255}, {0, 1.0, 2, 3, 4, 5, 6, 7, 520, 9.0, 10, 11.0, -250, 525, 270, 15, 16, -495, 530, 12, 20.0, 21, -505, 1275, 280, 25, 10001, 26.0, 24.0, 260, 30.0, 1055, 2550, 515, 35, 550, 1025, 300, 45.0, 560, 305, 50, 51, 52, 54.0, 55, 2555, 56.0, -195, 65.0, 2625, 75.0, 49996, 80, 4950, 605, 50015, 1375, 100, 102, 105, 110, 111, 625, 115, 4980, 120, -4996, 125.0, 130, 4996, 5510, 5255, -120, 5004, 5006, 5525, 150, 155.0, 5275, -100, 25001, 25005, 5555, 5125, 2755, -55, -54.0, 205, 2510, 3025, -45, -9.0, 2775, 55001, -550, -35.0, 2525, -30, -25, 1000, 1255, 250, 1001.0, -20.0, -24.0, 495, -15, -270, 5105, 1010, -10, -2.0, -520, 505, -6, -5, -4, -3, -1.0, 511}, -1, -1, -1, -1, -1]

'''

File: DP-1-expressN/expressN.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int N, int number)
{
    int answer = 0;
    if (number == N)
        return 1;
    vector<vector<int>> d(10, vector<int>(1, -1));
    d[0][0] = 0;
    d[1][0] = N;

    for (int i = 2; i < d.size(); i++)
    {
        vector<int> tmp;
        string s_N = "";
        for (int j = 0; j < i; j++)
            s_N += to_string(N);
        tmp.push_back(stoi(s_N));
        for (int j = 1; j <= i / 2; j++)
            for (int k = 0; k < d[j].size(); k++)
                for (int l = 0; l < d[i - j].size(); l++)
                {
                    tmp.push_back(d[j][k] * d[i - j][l]);
                    tmp.push_back(d[j][k] + d[i - j][l]);
                    if (d[i - j][l] != 0)
                        tmp.push_back(d[j][k] / d[i - j][l]);
                    if (d[j][k] != 0)
                        tmp.push_back(d[i - j][l] / d[j][k]);
                    tmp.push_back(d[j][k] - d[i - j][l]);
                    tmp.push_back(d[i - j][l] - d[j][k]);
                }
        tmp.erase(unique(tmp.begin(), tmp.end()), tmp.end());
        d[i] = tmp;
        for (int m = 0; m < d[i].size(); m++)
            if (d[i][m] == number)
                return i;
    }

    return -1;
}

int main(void)
{
    int N = 5, number = 12;

    cout << solution(N, number) << endl;

    return 0;
}